### PR TITLE
Firebase auth Email Action Links API

### PR DESCRIPTION
**Pull request by hardikns**
Added `generate_password_reset_link()`, 
  `generate_email_verification_link()` and `generate_email_sign_in_link()`
  methods to the `auth` API.

### CODE REVIEW

File: firebase_admin/_user_mgt.py [hunk 1]
```diff
@@ -372,6 +374,81 @@ def photo_url(self):
...
         return self._data.get('providerId')
 
+class ActionCodeSettings(object):
+    """Contains required continue/state URL with optional Android and iOS settings.
+    Used when invoking the email action link generation APIs.
+    """
+    KEYS = set(['url', 'handle_code_in_app', 'dynamic_link_domain', 'ios_bundle_id',
+                'android_package_name', 'android_minimum_version', 'android_install_app'])
+
+    def __init__(self, data=None):
```
Comment (hiranya911) [inline firebase_admin/_user_mgt.py]: This needs to accept a series of keyword arguments.

File: firebase_admin/_user_mgt.py [hunk 2]
```diff
@@ -372,6 +374,81 @@ def photo_url(self):
...
+        super(ActionCodeSettings, self).__init__()
+        data = data or {}
+        if not isinstance(data, dict):
+            raise ValueError('Invalid data argument: {0}. Must be a dictionary.'.format(data))
+        if len(set(six.iterkeys(data)) - self.KEYS):
+            raise ValueError('Invalid settings provided: {0}. Valid dictionary \
+                have following keys {1}'.format(data, self.KEYS))
+        self._data = copy.deepcopy(data)
+
+    @property
```
Comment (hiranya911) [inline firebase_admin/_user_mgt.py]: Define and assign attributes in the constructor. Follow the same pattern used in the FCM API in `_messaging_utils.py`.

File: firebase_admin/_auth_utils.py [hunk 3]
```diff
@@ -181,3 +182,74 @@ def validate_custom_claims(custom_claims, required=False):
...
+        raise ValueError('Invalid data argument: {0}. Must be a dictionary.'.format(settings))
+
+    parameters = {}
+    # Validate url
+    url = settings.get('url', None)
+    if url:
+        try:
+            parsed = urllib.parse.urlparse(url)
+            if not parsed.netloc:
+                raise ValueError('Malformed photo URL: "{0}".'.format(url))
```
Comment (hiranya911) [inline firebase_admin/_auth_utils.py]: Dynamic action links url

File: firebase_admin/_auth_utils.py [hunk 4]
```diff
@@ -181,3 +182,74 @@ def validate_custom_claims(custom_claims, required=False):
...
+    if android_package_name:
+        parameters['androidPackageName'] = android_package_name
+    if android_minimum_version:
+        parameters['androidMinimumVersion'] = android_minimum_version
+    if android_install_app:
+        parameters['androidInstallApp'] = android_install_app
+    return parameters
+
+def validate_action_type(action_type):
+    if not action_type in VALID_ACTION_TYPE:
```
Comment (hiranya911) [inline firebase_admin/_auth_utils.py]: `if action_type not in VALID_ACTION_TYPE:`

File: firebase_admin/_auth_utils.py [hunk 5]
```diff
@@ -181,3 +182,74 @@ def validate_custom_claims(custom_claims, required=False):
         raise ValueError(
             'Claim "{0}" is reserved, and must not be set.'.format(invalid_claims.pop()))
     return claims_str
+
+def validate_action_code_settings(settings):
```
Comment (hiranya911) [inline firebase_admin/_auth_utils.py]: Lets call this `encode_action_code_settings()` to be consistent with the terminology used in FCM. I would also consider moving this to the same file where ActionCodeSettings is defined.

File: firebase_admin/_user_mgt.py [hunk 6]
```diff
@@ -372,6 +374,81 @@ def photo_url(self):
...
+    def android_install_app(self):
+        return self._data.get('android_install_app', False)
+
+    @android_install_app.setter
+    def android_install_app(self, android_install_app):
+        self._data['android_install_app'] = android_install_app
+
+    def to_parameters_dict(self):
```
Comment (hiranya911) [inline firebase_admin/_user_mgt.py]: Lets remove this, and instead add an `encode_action_code_settings()` method to the module. It should accept an `ActionCodeSettings` instance, and return a `dict`.

File: firebase_admin/auth.py [hunk 7]
```diff
@@ -448,6 +453,75 @@ def import_users(users, hash_alg=None, app=None):
...
+        ValueError: If the provided arguments are invalid
+        AuthError: If an error occurs while generating the link
+    """
+    user_manager = _get_auth_service(app).user_manager
+    try:
+        return user_manager.generate_email_action_link('VERIFY_EMAIL', email, settings)
+    except _user_mgt.ApiCallError as error:
+        raise AuthError(error.code, str(error), error.detail)
+
+def generate_email_sign_in_link(email, settings=None, app=None):
```
Comment (hiranya911) [inline firebase_admin/auth.py]: generate_sign_in_with_email_link

Also settings are required here.

File: firebase_admin/_user_mgt.py [hunk 8]
```diff
@@ -372,6 +374,79 @@ def photo_url(self):
...
+        try:
+            parsed = urllib.parse.urlparse(settings.url)
+            if not parsed.netloc:
+                raise ValueError('Malformed dynamic action links url: "{0}".'.format(settings.url))
+            parameters['continueUrl'] = settings.url
+        except Exception:
+            raise ValueError('Malformed dynamic action links url: "{0}".'.format(settings.url))
+
+    # Validate boolean types
+    for field in ['handle_code_in_app', 'android_install_app']:
```
Comment (hiranya911) [inline firebase_admin/_user_mgt.py]: Let's be explicit and simple here.

```
if settings.handle_code_in_app is not None:
  if not isinstance(settings.handle_code_in_app, bool):
    raise (...)
  parameters['canHandleCodeInApp'] = settings.handle_code_in_app
```

Same for other optional parameters.

File: firebase_admin/_user_mgt.py [hunk 9]
```diff
@@ -537,6 +612,43 @@ def import_users(self, users, hash_alg=None):
...
+            ValueError: If the provided arguments are invalid
+        """
+        payload = {
+            'requestType': _auth_utils.validate_action_type(action_type),
+            'email': _auth_utils.validate_email(email),
+            'returnOobLink': True
+        }
+
+        if action_code_settings:
+            if not isinstance(action_code_settings, ActionCodeSettings):
```
Comment (hiranya911) [inline firebase_admin/_user_mgt.py]: We can remove this check.

File: firebase_admin/_user_mgt.py [hunk 10]
```diff
@@ -30,6 +31,7 @@
 USER_DELETE_ERROR = 'USER_DELETE_ERROR'
 USER_IMPORT_ERROR = 'USER_IMPORT_ERROR'
 USER_DOWNLOAD_ERROR = 'LIST_USERS_ERROR'
+USER_LINK_GENERATE_ERROR = 'USER_LINK_GENERATE_ERROR'
```
Comment (hiranya911) [inline firebase_admin/_user_mgt.py]: Lets call it GENERATE_EMAIL_ACTION_LINK_ERROR. Both name and value.

File: firebase_admin/_auth_utils.py [hunk 11]
```diff
@@ -181,3 +182,9 @@ def validate_custom_claims(custom_claims, required=False):
         raise ValueError(
             'Claim "{0}" is reserved, and must not be set.'.format(invalid_claims.pop()))
     return claims_str
+
+def validate_action_type(action_type):
+    if action_type not in VALID_ACTION_TYPE:
+        raise ValueError('Invalid action type provided action_type: {0}. \
+            Valid values are {1}'.format(action_type, VALID_ACTION_TYPE))
```
Comment (hiranya911) [inline firebase_admin/_auth_utils.py]: Pass `', '.join(VALID_ACTION_TYPES` to the format function.

File: firebase_admin/_auth_utils.py [hunk 12]
```diff
@@ -26,6 +26,7 @@
     'acr', 'amr', 'at_hash', 'aud', 'auth_time', 'azp', 'cnf', 'c_hash', 'exp', 'iat',
     'iss', 'jti', 'nbf', 'nonce', 'sub', 'firebase',
 ])
+VALID_ACTION_TYPE = set(['VERIFY_EMAIL', 'EMAIL_SIGNIN', 'PASSWORD_RESET'])
```
Comment (hiranya911) [inline firebase_admin/_auth_utils.py]: VALID_EMAIL_ACTION_TYPES

File: CHANGELOG.md [hunk 13]
```diff
@@ -1,5 +1,8 @@
 # Unreleased
 
+- [added] Added `generate_password_reset_link()`, 
+  `generate_email_verification_link()` and `generate_email_sign_in_link()`
```
Comment (hiranya911) [inline CHANGELOG.md]: generate_sign_in_with_email_link

File: firebase_admin/auth.py [hunk 14]
```diff
@@ -448,6 +453,78 @@ def import_users(users, hash_alg=None, app=None):
...
 
+def generate_password_reset_link(email, action_code_settings=None, app=None):
+    """Generates the out-of-band email action link for password reset flows for the specified email
+    address.
+
+    Args:
+        email: The email of the user whose password is to be reset.
+        action_code_settings: ``ActionCodeSettings`` instance (optional). Defines whether
+            the link is to be handled by a mobile app and the additional state information to be
+            passed in the deep link, etc.
```
Comment (hiranya911) [inline firebase_admin/auth.py]: Lets drop `etc` at the end.

File: firebase_admin/auth.py [hunk 15]
```diff
@@ -448,6 +453,78 @@ def import_users(users, hash_alg=None, app=None):
...
+
+def generate_email_verification_link(email, action_code_settings=None, app=None):
+    """Generates the out-of-band email action link for email verification flows for the specified
+    email address.
+
+    Args:
+        email: The email of the user to be verified.
+        action_code_settings: ``ActionCodeSettings`` instance (optional). Defines whether
+            the link is to be handled by a mobile app and the additional state information to be
+            passed in the deep link, etc.
```
Comment (hiranya911) [inline firebase_admin/auth.py]: Drop etc

File: firebase_admin/auth.py [hunk 16]
```diff
@@ -448,6 +453,78 @@ def import_users(users, hash_alg=None, app=None):
...
+
+def generate_sign_in_with_email_link(email, action_code_settings, app=None):
+    """Generates the out-of-band email action link for email link sign-in flows, using the action
+    code settings provided.
+
+    Args:
+        email: The email of the user signing in.
+        action_code_settings: ``ActionCodeSettings`` instance. Defines whether
+            the link is to be handled by a mobile app and the additional state information to be
+            passed in the deep link, etc.
```
Comment (hiranya911) [inline firebase_admin/auth.py]: Drop etc

File: integration/test_auth.py [hunk 17]
```diff
@@ -372,6 +373,37 @@ def test_import_users_with_password(api_key):
     finally:
         auth.delete_user(uid)
 
+@pytest.fixture
```
Comment (hiranya911) [inline integration/test_auth.py]: This object can be initialized in the test cases where it's needed. No need to turn it into a fixture.

File: integration/test_auth.py [hunk 18]
```diff
@@ -372,6 +373,37 @@ def test_import_users_with_password(api_key):
...
+def action_code_settings():
+    return auth.ActionCodeSettings('http://localhost')
+
+def _validate_link_url(link):
```
Comment (hiranya911) [inline integration/test_auth.py]: We need to do a stronger check here. See how it's done in Java for example.

https://github.com/firebase/firebase-admin-java/blob/master/src/test/java/com/google/firebase/auth/FirebaseAuthIT.java#L535

File: integration/test_auth.py [hunk 19]
```diff
@@ -372,6 +373,37 @@ def test_import_users_with_password(api_key):
...
+    assert isinstance(link, six.string_types)
+    six.moves.urllib.parse.urlparse(link)
+
+def test_password_reset(new_user_with_params):
+    link = auth.generate_password_reset_link(new_user_with_params.email)
+    _validate_link_url(link)
```
Comment (hiranya911) [inline integration/test_auth.py]: We should also probably do a password reset using the generated link and verify it works. If you're not sure how to do that, we can implement that bit later in a future PR.

File: tests/test_user_mgt.py [hunk 20]
```diff
@@ -972,3 +972,162 @@ def test_revoke_refresh_tokens(self, user_mgt_app):
...
+        'ios_bundle_id': 'test.bundle',
+        'android_package_name': 'test.bundle',
+        'android_minimum_version': '7',
+        'android_install_app': True,
+    }
+    return auth.ActionCodeSettings('http://localhost', **data)
+
+class TestGenerateEmailActionLink(object):
+
+    def _validate_request(self, request, settings=None):
```
Comment (hiranya911) [inline tests/test_user_mgt.py]: Move to the end of the class.

File: tests/test_user_mgt.py [hunk 21]
```diff
@@ -972,3 +972,162 @@ def test_revoke_refresh_tokens(self, user_mgt_app):
...
+def action_code_settings():
+    data = {
+        'handle_code_in_app': True,
+        'dynamic_link_domain': 'http://testly',
+        'ios_bundle_id': 'test.bundle',
+        'android_package_name': 'test.bundle',
+        'android_minimum_version': '7',
+        'android_install_app': True,
+    }
+    return auth.ActionCodeSettings('http://localhost', **data)
```
Comment (hiranya911) [inline tests/test_user_mgt.py]: Just make this a module-level constant.

File: tests/test_user_mgt.py [hunk 22]
```diff
@@ -972,3 +972,162 @@ def test_revoke_refresh_tokens(self, user_mgt_app):
...
+    def test_bad_settings_data(self, user_mgt_app, func):
+        _instrument_user_manager(user_mgt_app, 200, '{"oobLink":"https://testlink"}')
+        with pytest.raises(ValueError):
+            func("test@test.com", app=user_mgt_app, action_code_settings=1234)
+
+    def test_bad_action_type(self, user_mgt_app, action_code_settings):
+        with pytest.raises(ValueError):
+            auth._get_auth_service(user_mgt_app) \
+                .user_manager \
+                .generate_email_action_link("BAD_TYPE", "test@test.com",
```
Comment (hiranya911) [inline tests/test_user_mgt.py]: Single quote string literals. Here and everywhere else.

File: integration/test_auth.py (hunk 18 above)
Comment (hiranya911) [inline integration/test_auth.py]: Move this to the top where the other similar helper functions are defined.

**Comment by hardikns** [comment]
@hiranya911 , Thanks.  I will address these comments in couple of days.

File: integration/test_auth.py (hunk 17 above)
Comment (hardikns) [reply integration/test_auth.py]: Ok. may be my understanding is wrong - I thought its instantiated only when its injected in a method.

**Comment by hardikns** [comment]
@hiranya911 , I think I have fixed all comments, but will re-review tomorrow.

File: integration/test_auth.py [hunk 23]
```diff
@@ -373,37 +418,41 @@ def test_import_users_with_password(api_key):
...
-def test_email_verification(new_user_with_params):
-    link = auth.generate_email_verification_link(new_user_with_params.email)
-    _validate_link_url(link)
-
-def test_password_reset_with_settings(new_user_with_params, action_code_settings):
-    link = auth.generate_password_reset_link(new_user_with_params.email,
+def test_password_reset(new_user_email_unverified, api_key):
+    link = auth.generate_password_reset_link(new_user_email_unverified.email)
+    oob_code = _validate_link_url(link, check_continue_url=False)
+    assert new_user_email_unverified.email == _reset_password(oob_code, "newPassword", api_key)
```
Comment (hiranya911) [inline integration/test_auth.py]: Lets break this down into 2 lines for clarity:

```
user_email = _reset_password(oob_code, ...)
assert new_user_email_unverified.email == user_email
```

Here and below.

File: integration/test_auth.py [hunk 24]
```diff
@@ -55,6 +59,35 @@ def _random_id():
...
+    return resp.json().get('email')
+
+def _sign_in_with_email_link(email, oob_code, api_key):
+    body = {'oobCode': oob_code, 'email': email}
+    params = {'key' : api_key}
+    resp = requests.request('post', _email_sign_in_url, params=params, json=body)
+    resp.raise_for_status()
+    return resp.json().get('idToken')
+
+def _validate_link_url(link, check_continue_url=True):
```
Comment (hiranya911) [inline integration/test_auth.py]: Lets remove the assertion from this method, and return the entire `query_dict` from it. Then lets rename the method to `extract_link_params()` or something similar.

Callers can decide how to deal with the returned dict.

File: integration/test_auth.py (hunk 17 above)
Comment (hiranya911) [reply integration/test_auth.py]: It is, but this object is cheap to instantiate. Fixtures are best for heavy objects.

File: integration/test_auth.py (hunk 23 above)
Comment (hardikns) [reply integration/test_auth.py]: Done.

File: integration/test_auth.py (hunk 24 above)
Comment (hardikns) [reply integration/test_auth.py]: Done.

File: integration/test_auth.py [hunk 25]
```diff
@@ -420,38 +417,54 @@ def test_import_users_with_password(api_key):
 
 def test_password_reset(new_user_email_unverified, api_key):
     link = auth.generate_password_reset_link(new_user_email_unverified.email)
-    oob_code = _validate_link_url(link, check_continue_url=False)
-    assert new_user_email_unverified.email == _reset_password(oob_code, "newPassword", api_key)
+    assert isinstance(link, six.string_types)
+    query_dict = _extract_link_params(link)
+    user_email = _reset_password(query_dict['oobCode'], "newPassword", api_key)
```
Comment (hiranya911) [inline integration/test_auth.py]: Sorry I didn't notice this earlier. But please use single quotes for string literals: `'newPassword'`

File: integration/test_auth.py [hunk 26]
```diff
@@ -420,38 +417,54 @@ def test_import_users_with_password(api_key):
...
 def test_password_reset_with_settings(new_user_email_unverified, api_key):
     action_code_settings = auth.ActionCodeSettings(ACTION_LINK_CONTINUE_URL)
     link = auth.generate_password_reset_link(new_user_email_unverified.email,
                                              action_code_settings=action_code_settings)
-    oob_code = _validate_link_url(link)
-    assert new_user_email_unverified.email == _reset_password(oob_code, "newPassword", api_key)
+    assert isinstance(link, six.string_types)
+    query_dict = _extract_link_params(link)
+    assert query_dict['continueUrl'] == ACTION_LINK_CONTINUE_URL
+    user_email = _reset_password(query_dict['oobCode'], "newPassword", api_key)
```
Comment (hiranya911) [inline integration/test_auth.py]: Here too.

File: integration/test_auth.py [hunk 27]
```diff
@@ -420,38 +417,54 @@ def test_import_users_with_password(api_key):
...
     link = auth.generate_sign_in_with_email_link(new_user_email_unverified.email,
                                                  action_code_settings=action_code_settings)
-    oob_code = _validate_link_url(link)
-    assert _sign_in_with_email_link(new_user_email_unverified.email, oob_code, api_key)
+    assert isinstance(link, six.string_types)
+    query_dict = _extract_link_params(link)
+    assert query_dict['continueUrl'] == ACTION_LINK_CONTINUE_URL
+    oob_code = query_dict['oobCode']
+    id_token = _sign_in_with_email_link(new_user_email_unverified.email, oob_code, api_key)
+    assert id_token
```
Comment (hiranya911) [inline integration/test_auth.py]: Lets make this explicit: `assert id_token is not None and len(id_token) > 0`

File: integration/test_auth.py (hunk 25 above)
Comment (hardikns) [reply integration/test_auth.py]: done.

File: integration/test_auth.py (hunk 26 above)
Comment (hardikns) [reply integration/test_auth.py]: done

File: integration/test_auth.py (hunk 27 above)
Comment (hardikns) [reply integration/test_auth.py]: done

**Pull request merged**